d2/src/test/java/com/linkedin/d2/balancer/util/canary/TestBasicCanaryDistributionProviderImpl.java

```diff
@@ -0,0 +1,210 @@
+  public void testNormalCasesForApplicationsStrategy()
+  {
+    TargetApplicationsStrategyProperties appsProperties =
+        new TargetApplicationsStrategyProperties().setTargetApplications(new StringArray(Arrays.asList("appA", "appB")))
+            .setScope(0.4);
+    D2CanaryDistributionStrategy targetAppsStrategy =
+        new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_APPLICATIONS)
+            .setTargetApplicationsStrategyProperties(appsProperties);
+
+    CanaryDistributionProviderImplFixture fixture = new CanaryDistributionProviderImplFixture();
+
+    // NOT in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appC", null).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE, "App not in target list should return stable.");
+
+    // in scope and target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 38).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.CANARY,
+        "App in target list and hash result in canary scope should return canary.");
+
+    // not in scope and in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 50).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE,
+        "App in target list but hash result not in canary scope should return stable.");
+  }
+
+  /**
+   * Provide objects for testing edge cases
+   * @return a list of objects with structure: {
+   *  D2 canary distribution strategy
+   * }
+   */
+  @DataProvider(name = "getEdgeCaseStrategies")
+  public Object[][] getEdgeCaseStrategies()
+  {
+    return new Object[][]{
+        // percentage strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.PERCENTAGE)},
+        // target hosts strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_HOSTS)},
+        // target apps strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_APPLICATIONS)},
+        // strategy with disabled type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.DISABLED)},
+        // strategy with unknown type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.$UNKNOWN)}};
+  }
+
+  @Test(dataProvider = "getEdgeCaseStrategies")
+  public void testEdgeCases(D2CanaryDistributionStrategy strategy)
+  {
+    Assert.assertEquals(new CanaryDistributionProviderImplFixture().getSpiedImpl().distribute(strategy),
+        CanaryDistributionProvider.Distribution.STABLE, "Invalid strategies should return stable");
+  }
+
+  @Test
+  public void testHashDistribution()
+  {
+    // verify hash results are almost uniformly distributed
+    int bucketCount = 100;
+    int attemptCount = 1000;
+    int[] hits = new int[bucketCount]; // to count hits in each bucket 0 ~ 99.
+    String serviceName = "test-service";
+    String hostName = "lor1-app";
```

Comment:
I believe the host names that the service gets are fully qualified - this should probably test with that, or using both.

Reply:
could you specify what do you mean here? Do you mean to use the real host and service name that the test gets? We still need to append the string with the different attempt number, so it becomes xxx0, xxx1, xxx2, xxx3, ... xxx1000. It will be the same as using dummy service and host names, won't it?

Reply:
Yes, I mean using the full host name including the other domains. The distribution of the hashing algorithm could be different between something like lor1-app001 and lor1-app001.linkedin.prod. I think we should be validating with data as similar to production as possible.